File: src/solutions/java/N37SudokuSolver.java
package solutions.java;

public class N37SudokuSolver {

    public void solveSudoku(char[][] board) {
        doSolve(board);
    }

    public boolean doSolve(char[][] board) {
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[0].length; j++) {
                if (board[i][j] == '.') {
                    for (char c = '1'; c <= '9'; c++) {
                        if (isValid(board, i, j, c)) {
                            board[i][j] = c;
                            if (doSolve(board)) {
                                return true;
                            }
                            else {
                                board[i][j] = '.';
                            }
                        }
                    }
                    return false;
                }
            }
        }
        return true;
    }

    private boolean isValid(char[][] board, int row, int col, char num) {
        int partRow = (row / 3) * 3;
        int partCol = (col / 3) * 3;
        for (int i = 0; i < 9; i++) {
            if (board[i][col] == num || board[row][i] == num || board[partRow + i / 3][partCol + i % 3] == num) {
                return false;
            }
        }
        return true;
    }
}

File: src/solutions/scala/N37SudokuSolver.scala
package solutions.scala

object N37SudokuSolver {

  def solveSudoku(board: Array[Array[Char]]): Unit = {
    def possibleDigits(board: Array[Array[Char]], r: Int, c: Int): Seq[Char] = {
      def cells(i: Int) = Seq(board(r)(i), board(i)(c), board(3 * (r / 3) + i / 3)(3 * (c / 3) + i % 3))

      ('1' to '9').diff(board.indices.flatMap(cells))
    }

    def solve(board: Array[Array[Char]], cell: Int = 0): Option[Array[Array[Char]]] =
      (cell % 9, cell / 9) match {
        case (0, 9) => Some(board)
        case (r, c) if board(r)(c) != '.' => solve(board, cell + 1)
        case (r, c) =>
          possibleDigits(board, r, c)
            .flatMap(n => solve(board.updated(r, board(r).updated(c, n))))
            .headOption
      }

    val res = solve(board).head
    for {
      i <- 0 until 9
      j <- 0 until 9
    } yield board(i)(j) = res(i)(j)
  }
}
